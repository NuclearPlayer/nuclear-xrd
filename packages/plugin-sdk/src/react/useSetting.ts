import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
