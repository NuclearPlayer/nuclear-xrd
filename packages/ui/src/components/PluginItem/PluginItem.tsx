import { SettingsIcon, TriangleAlertIcon } from 'lucide-react';
import { FC, ReactNode } from 'react';

import { cn } from '../../utils';
import { Box } from '../Box';
import { Button } from '../Button';

type PluginItemProps = {
  name: string;
  author: string;
  description: string;
  icon?: ReactNode;
  onViewDetails?: () => void;
  className?: string;
  disabled?: boolean;
  warning?: boolean;
  warningText?: string;
  rightAccessory?: ReactNode;
};

export const PluginItem: FC<PluginItemProps> = ({
  name,
  author,
  description,
  icon,
  onViewDetails,
  className,
  disabled = false,
  warning = false,
  warningText,
  rightAccessory,
}) => {
  return (
    <Box
      variant="secondary"
      className={cn(
        {
          'ring-2 ring-accent-orange ring-inset': warning,
          'opacity-30': disabled,
        },
        'relative transition-opacity duration-250',
        className,
      )}
    >
      <div className={'flex flex-wrap items-start gap-4 w-full'}>
        {icon && (
          <Box
            variant="tertiary"
            shadow="none"
            className="flex-shrink-0 w-12 h-12 p-0 items-center justify-center overflow-hidden"
          >
            {icon}
          </Box>
        )}

        <div className="flex-1 min-w-0">
          <h3 className="font-bold text-foreground text-lg gap-4 leading-tight inline-flex items-center">
            {name}
            {onViewDetails && (
              <Button size="icon" onClick={onViewDetails} disabled={disabled}>
                <SettingsIcon size={20} />
              </Button>
            )}
          </h3>
          <p className="text-sm text-foreground-secondary mt-1">by {author}</p>
          <p className="text-sm text-foreground-secondary mt-2 leading-relaxed">
            {description}
          </p>
        </div>

        {/* TODO: Add a tooltip for the warning text */}
        {(warning || warningText) && (
          <div className="absolute -top-4 -left-2 flex items-center w-12 h-12">
            {warning && (
              <span className="inline-flex items-center justify-center bg-accent-orange text-black text-xs font-semibold rounded p-1 border-2 border-border">
                <TriangleAlertIcon className="fill-accent-yellow" />
              </span>
            )}
          </div>
        )}

        <div className="flex flex-col shrink-0 h-full sm:w-auto items-start justify-center sm:items-end">
          {rightAccessory}
        </div>
      </div>
    </Box>
  );
};
