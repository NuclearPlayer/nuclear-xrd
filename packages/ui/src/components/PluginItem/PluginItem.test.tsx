import { render } from '@testing-library/react';
import { Music } from 'lucide-react';

import { PluginItem } from './PluginItem';

describe('PluginItem', () => {
  it('(Snapshot) renders with all props', () => {
    const { asFragment } = render(
      <PluginItem
        className="border-accent-red"
        name="YouTube Music"
        author="Nuclear Team"
        description="Stream music directly from YouTube Music with full search and playlist support."
        icon={<Music size={24} />}
        onViewDetails={() => {}}
        loadTimeMs={200}
        warning
        warningText="Loaded with errors"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('(Snapshot) renders disabled state', () => {
    const { asFragment } = render(
      <PluginItem
        name="Disabled Plugin"
        author="Nuclear Team"
        description="This plugin is currently disabled."
        icon={<Music size={24} />}
        disabled
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('(Snapshot) renders rightAccessory content', () => {
    const { asFragment } = render(
      <PluginItem
        name="Accessory Plugin"
        author="Nuclear Team"
        description="Has accessory on the right."
        rightAccessory={<span>Accessory</span>}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
