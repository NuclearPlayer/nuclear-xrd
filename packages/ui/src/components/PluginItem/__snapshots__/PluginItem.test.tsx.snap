// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PluginItem > (Snapshot) renders disabled state 1`] = `
<DocumentFragment>
  <div
    class="border-border text-foreground flex h-full w-full rounded border-2 p-4 bg-background shadow-shadow opacity-30 relative cursor-default transition-opacity duration-250"
    data-testid="plugin-item"
  >
    <div
      class="flex w-full flex-wrap items-start gap-4"
    >
      <div
        class="border-border text-foreground flex rounded border-2 bg-background-secondary shadow-none h-12 w-12 flex-shrink-0 items-center justify-center overflow-hidden p-0"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-music"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 18V5l12-2v13"
          />
          <circle
            cx="6"
            cy="18"
            r="3"
          />
          <circle
            cx="18"
            cy="16"
            r="3"
          />
        </svg>
      </div>
      <div
        class="min-w-0 flex-1"
      >
        <h3
          class="text-foreground inline-flex items-center gap-4 text-lg leading-tight font-bold select-none"
          data-testid="plugin-name"
        >
          Disabled Plugin
        </h3>
        <p
          class="text-foreground-secondary mt-1 text-sm select-none"
          data-testid="plugin-author"
        >
          by Nuclear Team
        </p>
        <p
          class="text-foreground-secondary mt-2 text-sm leading-relaxed select-none"
          data-testid="plugin-description"
        >
          This plugin is currently disabled.
        </p>
      </div>
      <div
        class="flex h-full shrink-0 flex-col items-start justify-center sm:w-auto sm:items-end"
      />
      <span
        class="text-foreground-secondary absolute right-4 bottom-2 mt-2 text-sm"
      >
        Loaded in ms
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PluginItem > (Snapshot) renders rightAccessory content 1`] = `
<DocumentFragment>
  <div
    class="border-border text-foreground flex h-full w-full rounded border-2 p-4 bg-background shadow-shadow relative cursor-default transition-opacity duration-250"
    data-testid="plugin-item"
  >
    <div
      class="flex w-full flex-wrap items-start gap-4"
    >
      <div
        class="min-w-0 flex-1"
      >
        <h3
          class="text-foreground inline-flex items-center gap-4 text-lg leading-tight font-bold select-none"
          data-testid="plugin-name"
        >
          Accessory Plugin
        </h3>
        <p
          class="text-foreground-secondary mt-1 text-sm select-none"
          data-testid="plugin-author"
        >
          by Nuclear Team
        </p>
        <p
          class="text-foreground-secondary mt-2 text-sm leading-relaxed select-none"
          data-testid="plugin-description"
        >
          Has accessory on the right.
        </p>
      </div>
      <div
        class="flex h-full shrink-0 flex-col items-start justify-center sm:w-auto sm:items-end"
      >
        <span>
          Accessory
        </span>
      </div>
      <span
        class="text-foreground-secondary absolute right-4 bottom-2 mt-2 text-sm"
      >
        Loaded in ms
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PluginItem > (Snapshot) renders with all props 1`] = `
<DocumentFragment>
  <div
    class="text-foreground flex h-full w-full rounded border-2 p-4 bg-background shadow-shadow ring-accent-orange ring-2 select-none ring-inset relative cursor-default transition-opacity duration-250 border-accent-red"
    data-testid="plugin-item"
  >
    <div
      class="flex w-full flex-wrap items-start gap-4"
    >
      <div
        class="border-border text-foreground flex rounded border-2 bg-background-secondary shadow-none h-12 w-12 flex-shrink-0 items-center justify-center overflow-hidden p-0"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-music"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9 18V5l12-2v13"
          />
          <circle
            cx="6"
            cy="18"
            r="3"
          />
          <circle
            cx="18"
            cy="16"
            r="3"
          />
        </svg>
      </div>
      <div
        class="min-w-0 flex-1"
      >
        <h3
          class="text-foreground inline-flex items-center gap-4 text-lg leading-tight font-bold select-none"
          data-testid="plugin-name"
        >
          YouTube Music
          <button
            class="inline-flex cursor-pointer items-center rounded transition-all disabled:cursor-not-allowed disabled:opacity-50 text-foreground bg-primary border-border shadow-shadow hover:translate-x-shadow-x hover:translate-y-shadow-y border-2 hover:shadow-none size-10 justify-center"
            data-headlessui-state=""
            type="submit"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-settings"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.671 4.136a2.34 2.34 0 0 1 4.659 0 2.34 2.34 0 0 0 3.319 1.915 2.34 2.34 0 0 1 2.33 4.033 2.34 2.34 0 0 0 0 3.831 2.34 2.34 0 0 1-2.33 4.033 2.34 2.34 0 0 0-3.319 1.915 2.34 2.34 0 0 1-4.659 0 2.34 2.34 0 0 0-3.32-1.915 2.34 2.34 0 0 1-2.33-4.033 2.34 2.34 0 0 0 0-3.831A2.34 2.34 0 0 1 6.35 6.051a2.34 2.34 0 0 0 3.319-1.915"
              />
              <circle
                cx="12"
                cy="12"
                r="3"
              />
            </svg>
          </button>
        </h3>
        <p
          class="text-foreground-secondary mt-1 text-sm select-none"
          data-testid="plugin-author"
        >
          by Nuclear Team
        </p>
        <p
          class="text-foreground-secondary mt-2 text-sm leading-relaxed select-none"
          data-testid="plugin-description"
        >
          Stream music directly from YouTube Music with full search and playlist support.
        </p>
      </div>
      <div
        class="absolute -top-4 -left-2"
        data-headlessui-state=""
      >
        <button
          aria-expanded="false"
          class="cursor-pointer"
          data-headlessui-state=""
          id="headlessui-popover-button-:r2:"
          type="button"
        >
          <div
            class="relative flex h-12 w-12 items-center"
          >
            <span
              class="bg-accent-orange border-border inline-flex items-center justify-center rounded border-2 p-1 text-xs font-semibold text-black"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-triangle-alert fill-accent-yellow"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m21.73 18-8-14a2 2 0 0 0-3.48 0l-8 14A2 2 0 0 0 4 21h16a2 2 0 0 0 1.73-3"
                />
                <path
                  d="M12 9v4"
                />
                <path
                  d="M12 17h.01"
                />
              </svg>
            </span>
          </div>
        </button>
      </div>
      <div
        class="flex h-full shrink-0 flex-col items-start justify-center sm:w-auto sm:items-end"
      />
      <span
        class="text-foreground-secondary absolute right-4 bottom-2 mt-2 text-sm"
      >
        Loaded in 200ms
      </span>
    </div>
  </div>
</DocumentFragment>
`;
